src/components/views/rooms/EditMessageComposer.tsx

```diff
@@ -68,6 +69,48 @@ function getTextReplyFallback(mxEvent: MatrixEvent): string {
     return "";
 }
 
+/**
+ * Build the mentions information based on the editor model and the previous version
+ * of the event to generate a differential between the two:
+ *
+ * 1. Fetch mentioned users from the original event.
+ * 2. Search the model parts for room or user pills and fill in the mentions object
+ *    if the users or room were not mentioned in the original event.
+ *
+ * @param sender - The Matrix ID of the user sending the event.
+ * @param editedContent - The previous event content.
+ * @param model - The editor model to search for mentions, null if there is no editor.
+ */
+export function attachDifferentialMentions(sender: string, editedContent: IContent, model: EditorModel): IMentions {
+    // The mentions from the original event.
+    const prevMentions = editedContent["org.matrix.msc3952.mentions"];
+    const prevUserMentions = new Set();
+    if (Array.isArray(prevMentions?.user_ids)) {
+        prevMentions!.user_ids.forEach((userId) => prevUserMentions.add(userId));
+    }
+
+    // Generate the differential mentions: users who were not previously mentioned
+    // and a room mention if the previous event was not a room mention.
+    const mentions: IMentions = {};
+    const newUserMentions = new Set<string>();
+    for (const part of model.parts) {
+        if (part.type === Type.UserPill && !prevUserMentions.has(part.resourceId)) {
+            newUserMentions.add(part.resourceId);
+        } else if (part.type === Type.AtRoomPill && !prevMentions?.room) {
+            mentions.room = true;
+        }
+    }
+
+    // Ensure the *current* user isn't listed in the mentioned users.
+    newUserMentions.delete(sender);
+
+    if (newUserMentions.size) {
+        mentions.user_ids = [...newUserMentions];
+    }
+
+    return mentions;
+}
```

Comment:
This is largely similar to the `attachMentions` code - could we not just combine them and make the prev stuff optional?

Reply:
I attempted it at some point and it made the code quite confusing -- was focusing on correctness first. I'll take another look! :)